### PR TITLE
Refactoring backtesting sell conditions eval

### PR DESCRIPTION
**Pull request by jblestang**
## Summary

This PR aims is to share the same sell evaluation function among backtesting and the bot, so that when modifications are made to the bot, they get automagicaly in the bactesting  environnement. This PR also takes into account the fact that BUY cannot occur if SELL is on at the same time and that SELL cannot occur if BUY is on at the same time.

## Quick changelog

- refactoring bot to have a function returning True of False to indicate that the trade should be cloed
- refactoring backtesting to use this function
- refactoring backtesting to use dates instead of index

## Help required

Four unit tests are failing, but before modifying them I would like to have your point of view on what is failing and what is the best fix. The tests seems not very deterministic.

### CODE REVIEW

**Comment by kryofly** [comment]
Sorry for dumb question, but how can I test this out? I cant find a link to a patchset (a big diff file to apply) anywhere here in overbloated github. Can I git fetch it?

**Comment by glonlas** [comment]
@kryofly just add `.patch` at the end of the url.
https://patch-diff.githubusercontent.com/raw/gcarq/freqtrade/pull/450.patch

1. copy the patch into a file ex `patch.patch`
2. apply the patch `git apply patch.patch`

**Comment by kryofly** [comment]
Thanks @glonlas !  I wan't to check out those unit tests if they misbehave.

File: freqtrade/optimize/backtesting.py [hunk 1]
```diff
@@ -61,18 +61,15 @@ def generate_text_table(
...
-        results.profit.sum(),
-        results.loss.sum()
+        len(results[results.profit_BTC > 0]),
+        len(results[results.profit_BTC < 0])
     ])
     return tabulate(tabular_data, headers=headers, floatfmt=floatfmt)
 
 
-def get_trade_entry(pair, row, ticker, trade_count_lock, args):
+def get_sell_trade_entry(pair, row, rows, ticker, trade_count_lock, args):
```
Comment (kryofly) [inline freqtrade/optimize/backtesting.py]: Change rows to buy_subset ?

File: freqtrade/optimize/backtesting.py [hunk 2]
```diff
@@ -81,26 +78,20 @@ def get_trade_entry(pair, row, ticker, trade_count_lock, args):
...
-                          current_profit_btc,
-                          row2.Index - row.Index,
-                          current_profit_btc > 0,
-                          current_profit_btc < 0
-                          )
+                          trade.calc_profit_percent(rate=row2.close),
+                          trade.calc_profit(rate=row2.close),
+                          row2.Index - row.Index
+                          ), row2.date
+    return False
```
Comment (kryofly) [inline freqtrade/optimize/backtesting.py]: remove return False ?
Either we return a tuple or None, I guess it's more type-cleanier.

File: freqtrade/optimize/backtesting.py [hunk 3]
```diff
@@ -162,7 +153,7 @@ def backtest(args) -> DataFrame:
     if record and record.find('trades') >= 0:
         logger.info('Dumping backtest results')
         misc.file_dump_json('backtest-result.json', records)
-    labels = ['currency', 'profit_percent', 'profit_BTC', 'duration', 'profit', 'loss']
+    labels = ['currency', 'profit_percent', 'profit_BTC', 'duration']
```
Comment (kryofly) [inline freqtrade/optimize/backtesting.py]: Why the drop?

File: freqtrade/optimize/backtesting.py (hunk 1 above)
Comment (jblestang) [reply freqtrade/optimize/backtesting.py]: Yes, will do this modification

File: freqtrade/optimize/backtesting.py (hunk 3 above)
Comment (jblestang) [reply freqtrade/optimize/backtesting.py]: No need to get this done here, this is done in the function generating the  table with all results

File: freqtrade/optimize/backtesting.py (hunk 2 above)
Comment (jblestang) [reply freqtrade/optimize/backtesting.py]: OK for me!

File: freqtrade/main.py [hunk 4]
```diff
@@ -307,6 +307,29 @@ def min_roi_reached(trade: Trade, current_rate: float, current_time: datetime) -
     return False
 
 
+def should_sell(trade: Trade, rate: float, date: datetime, buy: bool, sell: bool) -> bool:
+    """
+    Sells the current pair if the threshold is reached and updates the trade record.
+    :return: True if trade has been sold, False otherwise
+    """
+    # Check if minimal roi has been reached and no longer in buy conditions (avoiding a fee)
+    if not buy and min_roi_reached(trade, rate, date):
```
[the remark this answers is not in the document]
Comment (jblestang) [reply freqtrade/main.py]: I'm feeling unconfortable with `if not buy and min_roi_reached(trade, rate, date):` and I'm wondering about not using the buy signal for this, this is a *I'm greedy person* optimisation

File: freqtrade/optimize/backtesting.py (hunk 3 above)
Comment (kryofly) [reply freqtrade/optimize/backtesting.py]: Cool! That would speed up hyperopt as well!

**Comment by jblestang** [comment]
Closing in favor of PR #469  (working branch is a mess , once more ...)

**Pull request closed without merging**